File: src/raycaster.ts
import { vec3 } from "gl-matrix";
import { World } from "./world/world";
import { BlockType } from "./world/block";

export interface RayHit {
  /** Integer world coordinates of the block that was hit. */
  bx: number;
  by: number;
  bz: number;
  /** Face normal (one of the 6 unit axis directions). */
  nx: number;
  ny: number;
  nz: number;
}

/**
 * DDA (Digital Differential Analyser) voxel ray traversal.
 *
 * Steps the ray one block boundary at a time – always crossing the nearest
 * axis-aligned boundary next – until it either hits a solid block or exceeds
 * `maxDist`.  This is O(n) in the number of blocks traversed and never
 * misses a block no matter how thin it is.
 *
 * Returns null if no solid block is within range.
 */
export function raycast(
  origin: vec3,
  direction: vec3,
  world: World,
  maxDist = 6,
): RayHit | null {
  // Normalise the direction just in case.
  const dir = vec3.normalize(vec3.create(), direction);

  // Current integer block position.
  let bx = Math.floor(origin[0]);
  let by = Math.floor(origin[1]);
  let bz = Math.floor(origin[2]);

  const dx = dir[0];
  const dy = dir[1];
  const dz = dir[2];

  // Step direction per axis (+1 or -1).
  const sx = dx >= 0 ? 1 : -1;
  const sy = dy >= 0 ? 1 : -1;
  const sz = dz >= 0 ? 1 : -1;

  // tDelta: how far along the ray (in t units) we must travel to cross one
  // block boundary on each axis.
  const tDeltaX = Math.abs(1 / dx);
  const tDeltaY = Math.abs(1 / dy);
  const tDeltaZ = Math.abs(1 / dz);

  // tMax: t value at which the ray crosses the next boundary on each axis.
  const tMaxX =
    dx === 0
      ? Infinity
      : (dx > 0 ? bx + 1 - origin[0] : origin[0] - bx) * tDeltaX;
  const tMaxY =
    dy === 0
      ? Infinity
      : (dy > 0 ? by + 1 - origin[1] : origin[1] - by) * tDeltaY;
  const tMaxZ =
    dz === 0
      ? Infinity
      : (dz > 0 ? bz + 1 - origin[2] : origin[2] - bz) * tDeltaZ;

  let tmx = tMaxX;
  let tmy = tMaxY;
  let tmz = tMaxZ;

  // Face normal of the last boundary crossed (updated each step).
  let nx = 0,
    ny = 0,
    nz = 0;

  while (Math.min(tmx, tmy, tmz) < maxDist) {
    // Advance to the nearest boundary.
    if (tmx < tmy && tmx < tmz) {
      bx += sx;
      nx = -sx;
      ny = 0;
      nz = 0;
      tmx += tDeltaX;
    } else if (tmy < tmz) {
      by += sy;
      nx = 0;
      ny = -sy;
      nz = 0;
      tmy += tDeltaY;
    } else {
      bz += sz;
      nx = 0;
      ny = 0;
      nz = -sz;
      tmz += tDeltaZ;
    }

    if (world.getBlock(bx, by, bz) !== BlockType.Air) {
      return { bx, by, bz, nx, ny, nz };
    }
  }

  return null;
}

File: game/src/engine/raycaster.rs
use glam::Vec3;

use crate::world::block::BlockType;
use crate::world::World;

pub struct RayHit {
    pub bx: i32,
    pub by: i32,
    pub bz: i32,
    /// Face normal (one of the 6 unit axis directions).
    pub nx: i32,
    pub ny: i32,
    pub nz: i32,
}

/// DDA (Digital Differential Analyser) voxel ray
/// traversal. Steps one block boundary at a time,
/// always crossing the nearest axis-aligned boundary
/// next. O(n) in blocks traversed, never misses.
pub fn raycast(
    origin: Vec3,
    direction: Vec3,
    world: &World,
    max_dist: f32,
) -> Option<RayHit> {
    let dir = direction.normalize();

    let mut bx = origin.x.floor() as i32;
    let mut by = origin.y.floor() as i32;
    let mut bz = origin.z.floor() as i32;

    let dx = dir.x;
    let dy = dir.y;
    let dz = dir.z;

    let sx = if dx >= 0.0 { 1i32 } else { -1 };
    let sy = if dy >= 0.0 { 1i32 } else { -1 };
    let sz = if dz >= 0.0 { 1i32 } else { -1 };

    let t_delta_x = (1.0 / dx).abs();
    let t_delta_y = (1.0 / dy).abs();
    let t_delta_z = (1.0 / dz).abs();

    let mut tmx = if dx == 0.0 {
        f32::INFINITY
    } else if dx > 0.0 {
        (bx as f32 + 1.0 - origin.x) * t_delta_x
    } else {
        (origin.x - bx as f32) * t_delta_x
    };
    let mut tmy = if dy == 0.0 {
        f32::INFINITY
    } else if dy > 0.0 {
        (by as f32 + 1.0 - origin.y) * t_delta_y
    } else {
        (origin.y - by as f32) * t_delta_y
    };
    let mut tmz = if dz == 0.0 {
        f32::INFINITY
    } else if dz > 0.0 {
        (bz as f32 + 1.0 - origin.z) * t_delta_z
    } else {
        (origin.z - bz as f32) * t_delta_z
    };

    let mut nx = 0i32;
    let mut ny = 0i32;
    let mut nz = 0i32;

    while tmx.min(tmy).min(tmz) < max_dist {
        if tmx < tmy && tmx < tmz {
            bx += sx;
            nx = -sx;
            ny = 0;
            nz = 0;
            tmx += t_delta_x;
        } else if tmy < tmz {
            by += sy;
            nx = 0;
            ny = -sy;
            nz = 0;
            tmy += t_delta_y;
        } else {
            bz += sz;
            nx = 0;
            ny = 0;
            nz = -sz;
            tmz += t_delta_z;
        }

        if world.get_block(bx, by, bz) != BlockType::Air
        {
            return Some(RayHit {
                bx,
                by,
                bz,
                nx,
                ny,
                nz,
            });
        }
    }

    None
}
